### ISSUE TITLE
[Build] GLX declaration conflicts (incompat with modern mesa headers?)

### ISSUE THREAD

**Issue by Houkime** [4ian/GDevelop#787]
## Describe the bug
Build fails with 

```
In file included from /home/user/.cache/aurman/gdevelop-git/src/GD/ExtLibs/SFML/src/SFML/Window/Unix/GlxContext.hpp:32,
  from /home/user/.cache/aurman/gdevelop-git/src/GD/ExtLibs/SFML/src/SFML/Window/GlContext.cpp:51:
/home/user/.cache/aurman/gdevelop-git/src/GD/ExtLibs/SFML/src/SFML/Window/Unix/GlxExtensions.hpp:123:3: error: conflicting declaration ‘typedef struct GLXHyperpipeNetworkSGIX GLXHyperpipeNetworkSGIX’
  } GLXHyperpipeNetworkSGIX;
      ^~~~~~~~~~~~~~~~~~~~~~~
In file included from /usr/include/GL/glx.h:333,
                                  from /home/user/.cache/aurman/gdevelop-git/src/GD/ExtLibs/SFML/src/SFML/Window/Unix/GlxExtensions.hpp:36,
                                  from /home/user/.cache/aurman/gdevelop-git/src/GD/ExtLibs/SFML/src/SFML/Window/Unix/GlxContext.hpp:32,
                                  from /home/user/.cache/aurman/gdevelop-git/src/GD/ExtLibs/SFML/src/SFML/Window/GlContext.cpp:51:
/usr/include/GL/glxext.h:764:3: note: previous declaration as ‘typedef struct GLXHyperpipeNetworkSGIX GLXHyperpipeNetworkSGIX’
  } GLXHyperpipeNetworkSGIX;
      ^~~~~~~~~~~~~~~~~~~~~~~
In file included from /home/user/.cache/aurman/gdevelop-git/src/GD/ExtLibs/SFML/src/SFML/Window/Unix/GlxContext.hpp:32,
                                  from /home/user/.cache/aurman/gdevelop-git/src/GD/ExtLibs/SFML/src/SFML/Window/GlContext.cpp:51:
/home/user/.cache/aurman/gdevelop-git/src/GD/ExtLibs/SFML/src/SFML/Window/Unix/GlxExtensions.hpp:129:3: error: conflicting declaration ‘typedef struct GLXHyperpipeConfigSGIX GLXHyperpipeConfigSGIX’
  } GLXHyperpipeConfigSGIX;
      ^~~~~~~~~~~~~~~~~~~~~~
In file included from /usr/include/GL/glx.h:333,
                                  from /home/user/.cache/aurman/gdevelop-git/src/GD/ExtLibs/SFML/src/SFML/Window/Unix/GlxExtensions.hpp:36,
                                  from /home/user/.cache/aurman/gdevelop-git/src/GD/ExtLibs/SFML/src/SFML/Window/Unix/GlxContext.hpp:32,
                                  from /home/user/.cache/aurman/gdevelop-git/src/GD/ExtLibs/SFML/src/SFML/Window/GlContext.cpp:51:
/usr/include/GL/glxext.h:770:3: note: previous declaration as ‘typedef struct GLXHyperpipeConfigSGIX GLXHyperpipeConfigSGIX’
  } GLXHyperpipeConfigSGIX;
      ^~~~~~~~~~~~~~~~~~~~~~
In file included from /home/user/.cache/aurman/gdevelop-git/src/GD/ExtLibs/SFML/src/SFML/Window/Unix/GlxContext.hpp:32,
                                  from /home/user/.cache/aurman/gdevelop-git/src/GD/ExtLibs/SFML/src/SFML/Window/GlContext.cpp:51:
/home/user/.cache/aurman/gdevelop-git/src/GD/ExtLibs/SFML/src/SFML/Window/Unix/GlxExtensions.hpp:134:3: error: conflicting declaration ‘typedef struct GLXPipeRect GLXPipeRect’
  } GLXPipeRect;
      ^~~~~~~~~~~
In file included from /usr/include/GL/glx.h:333,
                                  from /home/user/.cache/aurman/gdevelop-git/src/GD/ExtLibs/SFML/src/SFML/Window/Unix/GlxExtensions.hpp:36,
                                  from /home/user/.cache/aurman/gdevelop-git/src/GD/ExtLibs/SFML/src/SFML/Window/Unix/GlxContext.hpp:32,
                                  from /home/user/.cache/aurman/gdevelop-git/src/GD/ExtLibs/SFML/src/SFML/Window/GlContext.cpp:51:
/usr/include/GL/glxext.h:775:3: note: previous declaration as ‘typedef struct GLXPipeRect GLXPipeRect’
  } GLXPipeRect;
      ^~~~~~~~~~~
In file included from /home/user/.cache/aurman/gdevelop-git/src/GD/ExtLibs/SFML/src/SFML/Window/Unix/GlxContext.hpp:32,
                                  from /home/user/.cache/aurman/gdevelop-git/src/GD/ExtLibs/SFML/src/SFML/Window/GlContext.cpp:51:
/home/user/.cache/aurman/gdevelop-git/src/GD/ExtLibs/SFML/src/SFML/Window/Unix/GlxExtensions.hpp:138:3: error: conflicting declaration ‘typedef struct GLXPipeRectLimits GLXPipeRectLimits’
  } GLXPipeRectLimits;
      ^~~~~~~~~~~~~~~~~
In file included from /usr/include/GL/glx.h:333,
                                  from /home/user/.cache/aurman/gdevelop-git/src/GD/ExtLibs/SFML/src/SFML/Window/Unix/GlxExtensions.hpp:36,
                                  from /home/user/.cache/aurman/gdevelop-git/src/GD/ExtLibs/SFML/src/SFML/Window/Unix/GlxContext.hpp:32,
                                  from /home/user/.cache/aurman/gdevelop-git/src/GD/ExtLibs/SFML/src/SFML/Window/GlContext.cpp:51:
/usr/include/GL/glxext.h:779:3: note: previous declaration as ‘typedef struct GLXPipeRectLimits GLXPipeRectLimits’
  } GLXPipeRectLimits;
      ^~~~~~~~~~~~~~~~~
[  2%] Building CXX object ExtLibs/SFML/src/SFML/Audio/CMakeFiles/sfml-audio.dir/AudioDevice.cpp.o
make[2]: *** [ExtLibs/SFML/src/SFML/Window/CMakeFiles/sfml-window.dir/build.make:76: ExtLibs/SFML/src/SFML/Window/CMakeFiles/sfml-window.dir/GlContext.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:328: ExtLibs/SFML/src/SFML/Window/CMakeFiles/sfml-window.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....

```

## To Reproduce

* Try to compile (probably recent mesa needed)

## Other details

OS: Arch Linux (straightforward PKGBUILD without code patches)
mesa used (which provides glx.h) is master at 95d62baac5 (2 weeks ago)
https://gitlab.freedesktop.org/mesa/mesa/commit/95d62baac5daf39fc76bdffa600ffa2d5e36b25f

**Comment by 4ian** [comment]
Looks like an issue while compiling SFML (https://github.com/SFML/SFML).

Are you trying to compile GDevelop 4 or GDevelop.js for GDevelop 5?

**Comment by Houkime** [comment]
This is what comes in Cmake route, so this is GD4.
I've chosen this because I dislike node.js and electron which is mostly because such builds are rarely successful on my Arch, create too much (sometimes unneeded) files, make restrictions on them which constantly disrupt npm itself, fail/inconvenient to do stuff locally in the directory, and sometimes try to interfere with native pacman.
Electron is another far too often point of failure.

These systems just try to make too much and end up doing it too clumsy.
Cmake and c/c++ projects in general were far more reliable and easy to maintain in my experience.

**Comment by Houkime** [comment]
Hypothesis:

This problem wouldn't have existed if project used system-wide installed, always up to date sfml. However like in Godot you probably need to have static links to ensure portable executable at the end.
Thus you ended up having some static links and some dynamic (to mesa drivers?) which with different versions of software being included in the repository and installed on the system ended up in the conflict.

**Comment by 4ian** [comment]
> These systems just try to make too much and end up doing it too clumsy/convoluted.
> Cmake and c/c++ projects in general were far more reliable and easy to maintain in my experience.

I encourage you to take another look at GDevelop 5 :) In my experience of shipping GD4 and GD5, it's the contrary in terms of compatibility and maintainability.

Let me explain the architecture decision:

* GD5 is based on the same codebase as GDevelop 4 - the core library (Core and GDJS) is compiled from C++ to WebAssembly (well asm.js for now). 
* Electron is just there as a convenient shell on what could be a full Progressive Web App in the future. This allow extremely good portability with virtually 0 cross platform issue 
  * 👉A good example is that while GDevelop 4 was full of compatibility issue (basically it never worked properly on Linux without crashing randomly), GDevelop 5 runs on Windows/macOS/all recent Linux distros.
* Node.js is also here not used a lot. It offer access to the filesystem and it's almost all that it is used.
* It's actually hard to find good and customizable cross platform UI toolkit in the C++ world. On the contrary, JS frameworks and especially React in the recent years have brought an extremely powerful way of creating UI and even logic of any app using declarative components. This paradigm is really powerful and, according to my experience on 10 years building GDevelop, way better than the old way of creating UI with imperative APIs.
  * 👉A good example is Storybook. If you run `npm run storybook` (or `yarn storybook`), you are able to edit most components of the editor is isolation with **live reloading** in a browser. This is lightyears ahead of what you can do in C++.
* Having the whole editor in JavaScript makes it easy for new contributors to get involved - having Node.js installed is almost all you need.

In a nutshell, embracing some webtechnologies (WebGL and JavaScript) allow to put the burden of compatibility on browser vendors and benefit from good portability and an accessible code base, while keeping good performance.

I've made a talk about this architecture here: https://www.youtube.com/watch?v=6La7jSCnYyk

**Comment by 4ian** [comment]
Happy to merge a PR that updates SFML though! :) 
Hope the compatibility became better since then.

**Comment by blurymind** [comment]
I have to disagree about the point about node.js and electron - GD5 is actually easier to compile on windows/linux.
You are on Arch linux, so I assume that you might have the same problem I had with GD- not using the right version of nodejs.
Gd5 will fail to compile if you are on the latest version of nodejs  (CURRENT), which is a default on arch and manjaro. You need to use the older LTS version

Gd4 takes longer to compile, it's harder to set up and compile on windows. If you are using the buildscript on arch from aur, it's possible that you are compiling it with my script, that has been maintained by someone else now
https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=gdevelop-git

I think of creating a new script for gd5, which should make building it under arch just as easy

**Comment by 4ian** [comment]
> Gd5 will fail to compile if you are on the latest version of nodejs (CURRENT), which is a default on arch and manjaro. You need to use the older LTS version

Do you know why? I'm interested to see what's the reason behind this

**Comment by blurymind** [comment]
I remember googling the error and finding that it's nodejs related. Thus
why I closed it

**Comment by arthuro555** [comment]
This is an Old dead issue that doesn't relate to a bug in GDevelop itself, and only on a version with dropped support. I think this should be closed.
